### PR TITLE
Feature/import protocol upload

### PR DESCRIPTION
**Pull request by jabu404**
Protocol import and validation
Work done

- [x] Protocol upload using upload thing
- [x] Protocol upload button and drag and drop support
- [x] Protocol upload UI showing progress
- [x] Protocol download server side
- [ ] Protocol validation
- [x] Protocol assets upload using upload thing
- [ ] Protocol upload authorisation
- [x] useUploadThing hook

Screenshots:
Screenshot 2023-09-05 at 09 22 05
Screenshot 2023-09-05 at 12 13 52
Screenshot 2023-09-05 at 12 12 45
Screenshot 2023-09-05 at 12 12 18

### CODE REVIEW

**Review by buckhalt — CHANGES_REQUESTED** [review]
Overall, this is looking really great!

I think the UI of the upload status, loading states, and error/success messaging is solid. I have a few ideas for improving the UI of the import action where the user provides the file to import:
-  Adding a border to the area which will accept a file to be dropped to make the action more clear to the user.
- Have the existing Import Protocol button open a modal, similar to the existing component you created, which has the drop zone and button for opening local files to select.
- If possible, only make the button for browsing clickable, instead of the entire drop zone.

A rough example of what this could look like:
Screenshot 2023-09-05 at 12 55 58 PM

File: app/(main)/_actions/importProtocol.ts [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+  return protocol;
+};
+
+const MOCK_validateProtocol = async (_protocol: NCProtocol) => {
+  await sleep(100);
+  return {
+    dataErrors: [],
+    schemaErrors: [],
+  };
+  return {
```
[the remark this answers is not in the document]
Comment (jabu404) [reply app/(main)/_actions/importProtocol.ts:54]: Yeah this is just mocking validation, validation will be worked on in a separate pr

File: app/(main)/_actions/importProtocol.ts [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+import { hash } from 'bcrypt';
+import type Zip from 'jszip';
+import JSZip from 'jszip';
+import type { UploadFileResponse } from 'uploadthing/client';
+import type { FileEsque } from 'uploadthing/dist/sdk/utils';
+import { utapi } from 'uploadthing/server';
+import type {
+  AssetManifest,
+  Protocol as NCProtocol,
+} from '../../../lib/shared-consts';
```
[the remark this answers is not in the document]
Comment (jthrilly) [reply app/(main)/_actions/importProtocol.ts]: @buckhalt - yes, completely agree. 

@jabu404 - could you add that here?

**Review by jthrilly — COMMENTED** [review]
I think UI wise this is really great. You've done a brilliant job with the messaging and the progress bar.

Aside from the alias and the eslint rule to enforce it, the only other small UI nitpick is setting the correct cursor for the button (currently shows text select rather than pointer). 

I do agree with Caden about the dropzone UI state, but I know that styling that isn't a priority here. Perhaps just make the drop area a full width rectangle with some padding, and the button centered, and maybe add a border?

**Review by jthrilly — APPROVED** [review]
Looking great to me!

There are some changes in my user auth PR that might not merge cleanly with this, so I would prefer we get this in now so that I can rebase my work before I get too far into it.

Assuming you are happy with it, I am going to go ahead and merge now based on that. Any future changes can happen in a subsequent branch.

**Pull request merged**